trieste/acquisition/multi_objective/partition.py

```diff
@@ -64,46 +89,83 @@ def __post_init__(self) -> None:
+            For comparing results across runs, the same reference point must be used.
+        :return: lower, upper bounds of the partitioned cell
+        :raise ValueError (or `tf.errors.InvalidArgumentError`): If ``reference`` has an invalid
+            shape.
         """
-        tf.debugging.assert_rank(observations, 2)
-        tf.debugging.assert_greater_equal(tf.shape(observations)[-1], 2)
+        tf.debugging.assert_greater_equal(reference, self.front)
+        tf.debugging.assert_greater_equal(self.front, anti_reference)
+        tf.debugging.assert_type(anti_reference, self.front.dtype)
+        tf.debugging.assert_type(reference, self.front.dtype)
 
-        pfront, _ = non_dominated(observations)
-        self.front: Final[TensorType] = tf.gather_nd(pfront, tf.argsort(pfront[:, :1], axis=0))
-        self._bounds = self._get_bounds(self.front, jitter)
+        tf.debugging.assert_shapes(
+            [
+                (self._bounds.lower_idx, ["N", "D"]),
+                (self._bounds.upper_idx, ["N", "D"]),
+                (self.front, ["M", "D"]),
+                (reference, ["D"]),
+                (anti_reference, ["D"]),
+            ]
+        )
 
-    def _get_bounds(self, front: TensorType, jitter: float) -> _BoundedVolumes:
-        if front.shape[-1] > 2:
-            return self._divide_conquer_nd(front, jitter)
-        else:
-            return self._bounds_2d(front)
+        pseudo_pfront = tf.concat((anti_reference[None], self.front, reference[None]), axis=0)
+        N = tf.shape(self._bounds.upper_idx)[0]
+        D = tf.shape(self._bounds.upper_idx)[1]
+        idx = tf.tile(tf.range(D), (N,))
 
-    @staticmethod
-    def _bounds_2d(front: TensorType) -> _BoundedVolumes:
+        lower_idx = tf.stack((tf.reshape(self._bounds.lower_idx, [-1]), idx), axis=1)
+        upper_idx = tf.stack((tf.reshape(self._bounds.upper_idx, [-1]), idx), axis=1)
+
+        lower = tf.reshape(tf.gather_nd(pseudo_pfront, lower_idx), [N, D])
+        upper = tf.reshape(tf.gather_nd(pseudo_pfront, upper_idx), [N, D])
+
+        return lower, upper
+
+
+class ExactPartition2dNonDominated(BoundIndexPartition):
+    def __init__(self, front: TensorType):
+        """
+        :param front: non-dominated pareto front
+        """
+        tf.debugging.assert_equal(
+            tf.cast(tf.reduce_sum(tf.abs(non_dominated(front)[1])), dtype=front.dtype),
+            tf.zeros(shape=1, dtype=front.dtype),
+            message=f"\ninput {front} " f"contains dominated points",
+        )
+        self.front = tf.gather_nd(front, tf.argsort(front[:, :1], axis=0))  # sort input front
+        self._bounds = self._get_bound_index()
+
+    def _get_bound_index(self) -> _BoundedVolumes:
         # Compute the cells covering the non-dominated region for 2 dimension case
         # this assumes the Pareto set has been sorted in ascending order on the first
         # objective, which implies the second objective is sorted in descending order
```

Comment:
does 1st objective being ascending imply that the second objective is descending?

Reply:
yes, is the property of Pareto set (as is our input)

Reply:
perfect